From 65341361c5205553105993f20832ac0aa6e4ea1e Mon Sep 17 00:00:00 2001
From: Michal Orzel <michalx.orzel@intel.com>
Date: Thu, 22 Jun 2023 15:09:58 +0200
Subject: [PATCH] Add support for setting TLSv1.3 ciphers

Current version of library supports setting supported ciphers only for
TLSv1.2 and below. This patch adds possibility to set ciphersuites for
TLSv1.3 as well, by using new configuration option:
"ssl.tls13.ciphersuites".

Tested:
Verified that proper OpenSSL function for setting TLSv1.3 ciphersuites
is called.

Signed-off-by: Michal Orzel <michalx.orzel@intel.com>
---
 src/rdkafka_conf.c |  9 +++++++++
 src/rdkafka_conf.h |  1 +
 src/rdkafka_ssl.c  | 14 ++++++++++++++
 3 files changed, 24 insertions(+)

diff --git a/src/rdkafka_conf.c b/src/rdkafka_conf.c
index e481f4dd..e6542445 100644
--- a/src/rdkafka_conf.c
+++ b/src/rdkafka_conf.c
@@ -759,6 +759,15 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
              {RD_KAFKA_PROTO_SASL_SSL, "sasl_ssl", _UNSUPPORTED_SSL},
              {0, NULL}}},
 
+    {_RK_GLOBAL, "ssl.tls13.ciphersuites", _RK_C_STR,
+     _RK(ssl.tls13_ciphersuites),
+     "Set of ciphersuites supported by TLSv1.3. Due to major differences "
+     "between the way that ciphersuites for TLSv1.2 and below and "
+     "ciphersuites for TLSv1.3 work, they are configured in OpenSSL "
+     "differently too. The list should contain colon separated ciphersuite "
+     "names in order of preference. See manual page for "
+     "`SSL_CTX_set_cipher_list(3)`. OpenSSL >= 1.1.0 required.",
+     _UNSUPPORTED_OPENSSL_1_1_0},
     {_RK_GLOBAL, "ssl.cipher.suites", _RK_C_STR, _RK(ssl.cipher_suites),
      "A cipher suite is a named combination of authentication, "
      "encryption, MAC and key exchange algorithm used to negotiate the "
diff --git a/src/rdkafka_conf.h b/src/rdkafka_conf.h
index 161d6e46..ef2dbec5 100644
--- a/src/rdkafka_conf.h
+++ b/src/rdkafka_conf.h
@@ -229,6 +229,7 @@ struct rd_kafka_conf_s {
 #if WITH_SSL
                 SSL_CTX *ctx;
 #endif
+                char *tls13_ciphersuites;
                 char *cipher_suites;
                 char *curves_list;
                 char *sigalgs_list;
diff --git a/src/rdkafka_ssl.c b/src/rdkafka_ssl.c
index 9961a240..ac30b479 100644
--- a/src/rdkafka_ssl.c
+++ b/src/rdkafka_ssl.c
@@ -1668,6 +1668,20 @@ int rd_kafka_ssl_ctx_init(rd_kafka_t *rk, char *errstr, size_t errstr_size) {
         SSL_CTX_set_default_passwd_cb_userdata(ctx, rk);
 
         /* Ciphers */
+        if (rk->rk_conf.ssl.tls13_ciphersuites) {
+                rd_kafka_dbg(rk, SECURITY, "SSL",
+                             "Setting TLS1.3 ciphersuites: %s",
+                             rk->rk_conf.ssl.tls13_ciphersuites);
+                if (!SSL_CTX_set_ciphersuites(ctx,
+                                        rk->rk_conf.ssl.tls13_ciphersuites)) {
+                        /* Set a string that will prefix the
+                         * the OpenSSL error message (which is lousy)
+                         * to make it more meaningful. */
+                        rd_snprintf(errstr, errstr_size,
+                                    "ssl.tls13.ciphersuites failed: ");
+                        goto fail;
+                }
+        }
         if (rk->rk_conf.ssl.cipher_suites) {
                 rd_kafka_dbg(rk, SECURITY, "SSL", "Setting cipher list: %s",
                              rk->rk_conf.ssl.cipher_suites);
-- 
2.25.1

